drivers/misc/coresight/nrf_etr.c

```diff
@@ -30,23 +37,375 @@ LOG_MODULE_REGISTER(cs_etr_tbm);
+			fmt = "D%s\t%08x ";               \
+			break;                            \
+		default:                                  \
+			fmt = "D%s\t%016x ";              \
+			break;                            \
+		}                                         \
+		DBG(fmt, _marked ? "M" : "", _data);      \
+		for (int i = 0; i < _nlen / 2; i++) {     \
+			DBG("%c ", ((char *)&_data)[i]);  \
+		}                                         \
+		DBG("\n");                                \
+	} while (0)
+
 static const uint32_t wsize_mask = DT_REG_SIZE(ETR_BUFFER_NODE) / sizeof(int) - 1;
 static const uint32_t wsize_inc = DT_REG_SIZE(ETR_BUFFER_NODE) / sizeof(int) - 1;
 
+static bool in_sync;
+static int oosync_cnt;
 static volatile bool tbm_full;
 static volatile uint32_t base_wr_idx;
 static uint32_t etr_rd_idx;
+/* Counts number of new messages completed in the current formatter frame decoding. */
+static uint32_t new_msg_cnt;
+
 static bool volatile use_async_uart;
 
 static struct k_sem uart_sem;
 static const struct device *uart_dev = DEVICE_DT_GET(UART_NODE);
 static uint32_t frame_buf0[CORESIGHT_TRACE_FRAME_SIZE32] MEMORY_SECTION(UART_NODE);
 static uint32_t frame_buf1[CORESIGHT_TRACE_FRAME_SIZE32] MEMORY_SECTION(UART_NODE);
-static uint32_t *frame_buf = frame_buf0;
+static uint32_t frame_buf_decode[CORESIGHT_TRACE_FRAME_SIZE32];
+static uint32_t *frame_buf = IS_ENABLED(CONFIG_NRF_ETR_DECODE) ?
+				frame_buf_decode : frame_buf0;
 
 K_KERNEL_STACK_DEFINE(etr_stack, CONFIG_NRF_ETR_STACK_SIZE);
 static struct k_thread etr_thread;
 
+/* Domain details and prefixes. */
+static const uint16_t stm_m_id[] = {0x21, 0x22, 0x23, 0x2c, 0x2d, 0x2e, 0x24, 0x80};
+static const char *const stm_m_name[] = {"sec", "app", "rad", "sys", "flpr", "ppr", "mod", "hw"};
+static const char *const hw_evts[] = {
+	"CTI211_0",  /* 0 CTI211 triger out 1 */
+	"CTI211_1",  /* 1 CTI211 triger out 1 inverted */
+	"CTI211_2",  /* 2 CTI211 triger out 2 */
+	"CTI211_3",  /* 3 CTI211 triger out 2 inverted*/
+	"Sec up",    /* 4 Secure Domain up */
+	"Sec down",  /* 5 Secure Domain down */
+	"App up",    /* 6 Application Domain up */
+	"App down",  /* 7 Application Domain down */
+	"Rad up",    /* 8 Radio Domain up */
+	"Rad down",  /* 9 Radio Domain down */
+	"Radf up",   /* 10 Radio fast up */
+	"Radf down", /* 11 Radio fast down */
+	NULL, /* Reserved */
+	NULL, /* Reserved */
+	NULL, /* Reserved */
+	NULL, /* Reserved */
+	NULL, /* Reserved */
+	NULL, /* Reserved */
+	NULL, /* Reserved */
+	NULL, /* Reserved */
+	NULL, /* Reserved */
+	NULL, /* Reserved */
+	NULL, /* Reserved */
+	NULL, /* Reserved */
+	NULL, /* Reserved */
+	NULL, /* Reserved */
+	"GD LL up",    /* 26 Global domain low leakage up */
+	"GD LL down",  /* 27 Global domain low leakage down */
+	"GD1 HS up",   /* 28 Global domain high speed 1 up */
+	"GD1 HS up",   /* 29 Global domain high speed 1 up */
+	"GD0 HS down", /* 30 Global domain high speed 0 down */
+	"GD0 HS down", /* 31 Global domain high speed 0 down */
+};
```

Comment:
It seems this driver should be renamed to something like nrf54h_etr.c, because it includes soc-specific definitions. Alternatively, you could extract generic nrf_etr.c part and move soc-specific definitions to a soc-specific driver using the generic one.

Reply:
It is intended to be used for all nrf chips which have Coresight/ETR so i don't think that renaming it to `nrf54h_etr.c` would be good.

Reply:
Ok, I'm probably thinking about potential hardware enhancements that are not planned. So I guess we can leave it as is and potentially refactor when required